### DISCUSSION TITLE
leaf-list diff

### DISCUSSION THREAD

**Discussion by pkmuthu83** [CESNET/libyang#2235] [Q&A]
The diff API does not seem to give full diffs/annotations while changing order of elements in leaf-list. For ex, if we have leaf-list "testleaf", and the data tree dtree1 has the value [A, B], while the dtree2 has [B, A]. When we do dtree1.diff(dtree2), we get :

             {
            "testleaf": [
              "B"
            ],
            "@testleaf": [
              {
                "yang:operation": "replace",
                "yang:orig-default": false,
                "yang:orig-value": "A",
                "yang:value": ""
              }
            ]
          }

This does not give full idea of the diff, right? The first element A changed B, and second element B changed A.
Shouldn't it be something like below:
             {
            "testleaf": [
              "B",
              "A",
            ],
            "@testleaf": [
              {
                "yang:operation": "replace",
                "yang:orig-default": false,
                "yang:orig-value": "A",
                "yang:value": ""
              },
             {
                "yang:operation": "replace",
                "yang:orig-default": false,
                "yang:orig-value": "B",
                "yang:value": ""
              }
            ]
          }

**Comment by michalvasko** [comment]
Well, the term `diff` may be misleading, I suppose. The meaning of diff between data tree A and B is a data tree with operations that, when applied to data tree A, you get data tree B.

**Comment by pkmuthu83** [comment]
Thanks for responding. But, can you please elaborate on how the below diff can be applied to dtree1 to get dtree2 ?
```json
         {
        "testleaf": [
          "B"
        ],
        "@testleaf": [
          {
            "yang:operation": "replace",
            "yang:orig-default": false,
            "yang:orig-value": "A",
            "yang:value": ""
          }
        ]
      }
```

**Reply by michalvasko** [reply]
In short, call `lyd_diff_apply_all()`.
